### ISSUE TITLE
Speeding up the creation of a nc_benchmark

### ISSUE THREAD

**Issue by niniack** [ContinualAI/avalanche#1335]
Hello, 

I'm using a custom dataset to create a `nc_benchmark`. The dataset is fairly large, so it takes a while. Still,  I would like to speed this up.

I looked through the docs for `nc_benchmark` to see if I could provide any arguments to use more workers somehow, but I haven't found anything. I mention this because I noticed that I use a single core when creating a benchmark.

Here's how I'm loading my benchmark:

In a file called `benchmark.py`:

```python
def SplitCustom(
    n_experiences: int,
    root: str,
    *,
    return_task_id=False,
    seed: Optional[int] = None,
    fixed_class_order: Optional[Sequence[int]] = None,
    shuffle: bool = False,
    class_ids_from_zero_in_each_exp: bool = False,
    train_transform: Optional[Any] = None,
    eval_transform: Optional[Any] = None,
    dataset_root: Union[str, Path] = None,
) -> NCScenario:

    custom_train = CustomDataset(root=root, split="train")
    custom_test = CustomDataset(root=root, split="test")

    return nc_benchmark(
        train_dataset=custom_train,
        test_dataset=custom_test,
        n_experiences=n_experiences,
        task_labels=return_task_id,
        seed=seed,
        fixed_class_order=fixed_class_order,
        shuffle=shuffle,
        per_exp_classes=None,
        class_ids_from_zero_in_each_exp=class_ids_from_zero_in_each_exp,
        train_transform=train_transform,
        eval_transform=eval_transform,
    )
```

and in another file `dataset.py`:
```python
class CustomDataset(Dataset):
    def __init__(self,
                 root: str,
                 split: str,
                 transform: Optional[Callable] = None,
                 ):
        self.root = root
        self.split = split

        self.split_dir = "custom_" + split + "/"
        self.full_path = Path(self.root).joinpath(self.split_dir)
        self.files = list(self.full_path.glob('*.npz'))

        self.transform = transform

        # preload targets for NC benchmark
        target_path = self.full_path.joinpath('targets.pkl')
        with open(target_path, 'rb') as file:
            self.targets = pickle.load(file)

    def __len__(self):
        """
        Returns the size of the dataset
        """

        return len(self.targets)

    def __getitem__(self, idx):
        np_img = np.load(str(self.files[idx]))['image']
        np_label = np.load(str(self.files[idx]))['label']
        np_img = preprocess_input(np_img)

        image = torch.from_numpy(np.transpose(np_img, (2, 0, 1))).float()
        label = int(np_label)

        if self.transform is not None:
            image = self.transform(image)

        return image, label

```

Finally, in my `train.py` file:
```python
split_custom = SplitCustom(
        n_experiences=num_experiences,
        root="/path/to/custom/dataset")
train_stream = split_custom.train_stream
test_stream = split_custom.test_stream
```

It seems the `nc_benchmark` method is what is taking a while. Is there a more clever way to go about this? I would like to have a train/test stream where I can control which classes are in a single experience.

**Comment by AntonioCarta** [comment]
This is unexpected, but we don't have many tests with large datasets. How large is it? In theory, as long as the targets can be manipulated efficiently there should be no problem. @lrzpellegrini do you have any idea of what's happening?

**Comment by lrzpellegrini** [comment]
Hi @niniack, I'll look into this! 

I need to know:
- How many elements are there in the dataset and how many classes
- How many experiences you are trying to create
- The type of the targets field
- The version of Avalanche

**Comment by niniack** [comment]
@AntonioCarta, thanks for picking this up.

Hi @lrzpellegrini, thanks for your response and help. I'll try to be as detailed as possible. 

My dataset has a train/val/test split. To save myself time as I develop my strategy, I have been working with only the val and test directories. I pretend my val set is my train set. 

> How many elements are there in the dataset and how many classes

The dataset is actually just a subset of Imagenet-1K. So, there are 1k classes. The val set has 17679 elements. The test set has 17407 elements.

> How many experiences you are trying to create

So far I have been doing 1 class per experience. 1000 experiences.

> The type of the targets field

When defining the targets in the Dataset class, this is what I do.

```
self.targets = [
                int(np.load(str(self.files[x]))['label']) for x in range(len(self.files))
            ]
```

It's just a list of ints, where each int is a value between 1-1000 corresponding to the class each image is.

> The version of Avalanche

0.3.1

In effect, I'm just working with a smaller Imagenet. Each class is present in this custom dataset.

Your question made me realize that the problem varies with the number of experiences I try to create.

I know it is also affected by the number of elements in the dataset because I created a "fake" dataset where I picked one image from each class, resulting in 1k elements for both train and test. Then, I make a `nc_benchmark` of 1k experiences. This is relatively faster. 

I'm not sure how helpful this will be, but here are some stats. The 1k elements refers to the "fake" dataset I describe above. The ~17k elements refers to the validation set and test set. As described earlier, they're both similar in size.

| # Elements PER train/test stream      | # Experiences |  Time (s)|
| ----------- | ----------- | ----------- |
|  1k      | 1       | 0.378 |
|  1k      | 50       | 0.425 |
|  1k      | 100       | 0.442 |
|  1k      | 500       | 0.908 |
|  1k      | 1000       | 1.330 |
|  ~17k  | 1       | 0.13 |
|  ~17k  | 50       | 0.903 |
|  ~17k  | 100       | 1.486 |
|  ~17k  | 500       | 7.056 |
|  ~17k  | 1000       |13.84 |

Is this sort of scaling expected? Am I just spoiled? :) 

I apologize, I'm not familiar with what is happening under the hood when creating a `nc_benchmark` but I wonder if there is something parallelizable?

Please let me know if there is additional information I can provide!

**Comment by niniack** [comment]
I should add: 

I recognized that loading all the targets by iterating through the files in the val and test directories could be very slow. I pickled the targets and read the pickled file, so no time is wasted on the list comprehension I showed above. The experiments account for this!

**Comment by niniack** [comment]
Hi, some updates: 


I isolated the expensive part to [this chunk of code](https://github.com/ContinualAI/avalanche/blob/a92fe8dde2b121cfd630ca20fb8385dd6882115f/avalanche/benchmarks/scenarios/new_classes/nc_scenario.py#L396) in `nc_scenario.py` (Note: i'm only showing the training dataset part but its the same code chunk for test dataset): 

```python
for idx, element in enumerate(original_training_dataset.targets):
	if element in selected_classes:
		selected_indexes_train.append(idx)
		if  self._has_task_labels:
			pattern_train_task_labels[idx] = exp_id
```

(excuse the poor pseudocode that follows) 
Currently:
```
selected_indices <= []
class_set <= classes we want in this experience
for each target in targets:
	is the target in class_set?
		APPEND index of target to selected_indices
```

I propose:
```

selected_indices <= []
class_set <= classes we want in this experience
for each class in class_set:
	class_indices <= lookup indices of class from targets `val_to_idx` dict
	EXTEND selected_indices with class_indices
```
Way faster! I did the brave task of testing on my training dataset (not the measly val set anymore). The train set consists of 325644 elements. 

Creating a 1000 experiences with the existing method costs ~132 seconds. With the new method, it costs ~2 seconds.

How I'm achieving my suggested pseudocode:
```python
for sc in selected_classes:
    selected_indexes_train.extend(
	    original_training_dataset.targets.val_to_idx[sc]
	)
```

Of course, there is the glaring issue that I just totally didn't account for the logic that updates the `pattern_train_task_labels`, but baby steps. 

I wanted to post this to ask if this is meaningful at all? Maybe we can iterate on this? Excited to help!! I was excited to find the `val_to_idx` structure in the `targets` object.

**Comment by AntonioCarta** [comment]
Thanks for finding this out! As I told you, there is no reason why the generator should this slow. I have to look more in detail, but I believe that your solution is correct. If you can make a draft PR we can look into it more in detail. Benchmarks are well tested so any problem should be caught by the unit tests.

**Comment by niniack** [comment]
Sweet, glad you agree! I'll open up a draft PR in a while, just wrangling with some other deadlines at the moment. I also want to tackle the `pattern_train_task_labels` problem I described. I have not put any thought into that yet. 

 Feel free to assign this issue to me, if that's appropriate.

**Comment by niniack** [comment]
Created a PR.

To fill the `pattern_*_task_labels`, I just loop through all the targets we grab and fill the array for each target. Doesn't seem to harm performance.
